### PR TITLE
QUIC: Test malformed TPARAMs

### PR DESCRIPTION
**Pull request by hlandau**
Fixes openssl/project#93

### CODE REVIEW

**Review by t8m — APPROVED** [review]
I would say that the test fits better with the purpose of quicfaultstest.c but that is just a bikeshedding so approved without moving it anyway.

File: test/quicapitest.c [hunk 1]
```diff
@@ -1553,6 +1554,521 @@ static int test_noisy_dgram(int idx)
...
+    if (!TEST_true(WPACKET_get_total_written(&wpkt, &written)))
+        goto err;
+
+    WPACKET_finish(&wpkt);
+    have_wpkt = 0;
+
+    /*
+     * Append the constructed extension blob to the extension block.
+     */
+    old_len = ee->extensionslen;
```
Comment (mattcaswell) [inline test/quicapitest.c:1982]: Side note: This append to the end approach won't work if we ever extend this test to include session resumption. The psk TLS extension *must* always come last.

File: test/quicapitest.c [hunk 2]
```diff
@@ -1553,6 +1554,521 @@ static int test_noisy_dgram(int idx)
...
+
+    if (!qtest_fault_resize_message(qtf, ee->extensionslen + written))
+        goto err;
+
+    memcpy(ee->extensions + old_len, new_bufm->data, written);
```
Comment (mattcaswell) [inline test/quicapitest.c:1987]: Side note: we maybe should have a `qtest_fault_add_extension`. Ok for this for now - but I wouldn't want to repeat this code in other tests.

File: ssl/quic/quic_channel.c [hunk 3]
```diff
@@ -1283,8 +1283,10 @@ static int ch_on_transport_params(const unsigned char *params,
     QUIC_CONN_ID cid;
     const char *reason = "bad transport parameter";
 
-    if (ch->got_remote_transport_params)
+    if (ch->got_remote_transport_params) {
+        reason = "multiple transport parameter extensions";
```
Comment (mattcaswell) [inline ssl/quic/quic_channel.c:1287]: Can this ever actually happen? I would have expected this to be blocked at the TLS level...but still ok to have this check.

**Review by mattcaswell — APPROVED** [review]
LGTM. A few side notes - but nothing preventing approval.

File: ssl/quic/quic_channel.c (hunk 3 above)
Comment (hlandau) [reply ssl/quic/quic_channel.c:1287]: You'd know better than me &mdash; I assumed the same TLS extension can appear multiple times in a handshake.

File: test/quicapitest.c (hunk 1 above)
Comment (hlandau) [reply test/quicapitest.c:1982]: Interesting &mdash; noted.

File: test/quicapitest.c (hunk 2 above)
Comment (hlandau) [reply test/quicapitest.c:1987]: Yeah, it's definitely a gap.

**Comment by t8m** [comment]
Merged to the master and 3.2 branches. Thank you.

**Pull request closed without merging**